File: 684-redundant-connection/redundant-connection.java
class DSU {

    int[] parent, rank;
    public DSU(int n) {
        this.parent = new int[n];
        this.rank = new int[n];
        for(int i = 0; i < n; i++) {
            parent[i] = i;
        }
    }

    public int find(int x) {
        if(parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    public void  union(int x, int y)
    {
        int parentX = find(x);
        int parentY = find(y);

        if(parentX == parentY){
            return;
        }

        if(rank[parentX] == rank[parentY]){
            parent[parentX] = parentY;
            rank[parentY]++;
        }else if(rank[parentX] < rank[parentY]) {
            parent[parentX] = parentY;
        }
        else {
            parent[parentY] = parentX;
        }
    }
}

class Solution {
    public int[] findRedundantConnection(int[][] edges) 
    {
        int n = edges.length;
        // graph is 1 based
        DSU dsu = new DSU(n + 1);

        for(int[] edge : edges)
        {
            int x = edge[0];
            int y = edge[1];

            if(dsu.find(x) == dsu.find(y))
            return edge;

            dsu.union(x, y);
        }

        // No redundant edge found
        return new int[0]; 
    }
}

File: 684-redundant-connection/redundant-connection.cpp
class Solution {
public:
    
    class DSU {
    public:

        vector<int> parent, rank;
        DSU(int n)
        {
            for(int i = 0; i < n; i++)
            {
                parent.push_back(i);
                rank.push_back(0);
            }
        }

        int find(int x) 
        {
            if(x == parent[x])
            return x;
            
            // Path compression
            return parent[x] = find(parent[x]);
        }

        void unionSet(int x, int y) 
        {
            int parentX = find(x);
            int parentY = find(y);
            
            if(parentX == parentY)
            return;

            // Union by rank
            // rank same - anyone can be made parent
            if(rank[parentX] == rank[parentY]) 
            {
                parent[parentX] = parentY;
                rank[parentY]++;
            }

            // rank of parent of 'x' is higher
            else if(rank[parentX] > rank[parentY])
            parent[parentY] = parentX;

            // rank of parent of 'y' is higher
            else
            parent[parentX] = parentY;

        }
    };
    

    vector<int> findRedundantConnection(vector<vector<int>>& edges) 
    {
        int N = edges.size();
        DSU* dsu = new DSU(N);

        for(auto edge : edges)
        {
            int x = edge[0] - 1;
            int y = edge[1] - 1;

            if(dsu->find(x) == dsu->find(y))
            return edge;

            dsu->unionSet(x, y);
        }
        return {};
    }
};